### PR TITLE
#[UIC-1321]: Implement filters publication and subscription.

### PR DESCRIPTION
**Pull request by ankursinghal2005**
### CATEGORY
[UIC-1321](https://guavus-jira.atlassian.net/browse/UIC-1321)

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [x] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Implement filters publication and subscription.
We have added all operator support.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/assets/src/dashboard/util/charts/getEffectiveExtraFilters.js [hunk 1]
```diff
@@ -47,11 +75,26 @@ export default function getEffectiveExtraFilters({
     const filtersFromSlice = filters[filteringSliceId];
     Object.keys(filtersFromSlice).forEach(field => {
       if (!immuneToFields.includes(field)) {
-        effectiveFilters.push({
-          col: field,
-          op: 'in',
-          val: filtersFromSlice[field],
-        });
```
Comment (jitendra-kumawat) [inline superset/assets/src/dashboard/util/charts/getEffectiveExtraFilters.js]: why we remove immuneToFields logic to add filters in immuneToFields

File: superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js [hunk 2]
```diff
@@ -74,7 +110,9 @@ export default function getFormDataWithExtraFilters({
     extra_filters: getEffectiveExtraFilters({
       dashboardMetadata,
       filters,
-      sliceId
+      sliceId,
+      filterKeys,
+      linkedSlicesExistInFilters,
```
Comment (jitendra-kumawat) [inline superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js]: if filterKeys is  const list of values then we can skip it as args and refer as const vis import
rename it as TIME_FILTER_KEYS

File: superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js [hunk 3]
```diff
@@ -23,42 +23,78 @@ import getEffectiveExtraFilters from './getEffectiveExtraFilters';
...
     }
-  }
-  return slicesInState;
+  }  
+  return linkedSlicesExistInFilters;
+}
+
+const getSliceSubscriberMap = (publishSubscriberMap, sliceId) => {
+  return publishSubscriberMap && publishSubscriberMap.hasOwnProperty("subscribers") && publishSubscriberMap.subscribers && publishSubscriberMap.subscribers[sliceId] ? publishSubscriberMap.subscribers[sliceId]: undefined;
 }
```
Comment (jitendra-kumawat) [inline superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js]: property check miss for sliceId in  publishSubscriberMap.subscribers[sliceId]

File: superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js [hunk 4]
```diff
@@ -23,42 +23,78 @@ import getEffectiveExtraFilters from './getEffectiveExtraFilters';
...
+      }
     }
-  }
-  return slicesInState;
+  }  
+  return linkedSlicesExistInFilters;
+}
+
+const getSliceSubscriberMap = (publishSubscriberMap, sliceId) => {
+  return publishSubscriberMap && publishSubscriberMap.hasOwnProperty("subscribers") && publishSubscriberMap.subscribers && publishSubscriberMap.subscribers[sliceId] ? publishSubscriberMap.subscribers[sliceId]: undefined;
```
Comment (jitendra-kumawat) [inline superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js]: rename getSliceSubscriberMap to getSubscriberSlice

File: superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js [hunk 5]
```diff
@@ -23,42 +23,78 @@ import getEffectiveExtraFilters from './getEffectiveExtraFilters';
...
 
-const getLinkedSliceWithFilters = (formData, globalFilters) => {
-  let slicesInState = [];
-  const propExist = formData.hasOwnProperty("linked_slice");
-  if(propExist) {
-    const linked_slices = formData.linked_slice;
-    if (linked_slices instanceof Array) {
-      slicesInState  = linked_slices.filter(element => globalFilters.hasOwnProperty(element));  
+const getLinkedSlicesExistInFilters = (subscriberMap, globalFilters) => {
+  let linkedSlicesExistInFilters = [];
```
Comment (jitendra-kumawat) [inline superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js]: args name must match with pass args name

Comment (jitendra-kumawat) [inline superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js]: can we rename getLinkedSlicesExistInFilters to getPubSlicesOfSubscribeSlicesMatchesInFilters

File: superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js [hunk 6]
```diff
@@ -23,42 +23,78 @@ import getEffectiveExtraFilters from './getEffectiveExtraFilters';
...
+    if (subscriberMap && subscriberMap.actions.indexOf("APPLY_FILTER") > -1) {
+      const filters = getFiltersFromSlices(slicesInState,globalFilters)
+      return filters;
+    }
+    return {};
 }
-
+  
 const getFiltersFromSlices = (slices, globalFilters) => {
   const filters = {};
```
Comment (jitendra-kumawat) [inline superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js]: rename getFiltersFromSlices to createFiltersOfSubscriberColumnsInSlice

File: superset/assets/src/dashboard/reducers/getInitialState.js [hunk 7]
```diff
@@ -67,34 +67,52 @@ export default function(bootstrapData) {
...
+      }
+    })
+    return slice_data;
+  }
+
+  try {
+    let publishers = publishSubscriberMap.hasOwnProperty("publishers")? publishSubscriberMap["publishers"]:{};
+    for (var publish_id in publishers) {
+      if (publishers.hasOwnProperty(publish_id)) {
+
```
[the remark this answers is not in the document]
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/reducers/getInitialState.js]: done

File: superset/assets/src/dashboard/reducers/getInitialState.js [hunk 8]
```diff
@@ -67,34 +67,52 @@ export default function(bootstrapData) {
...
+          publish_columns.forEach( col => {
+            if (isPublishColumnExistsInFilters(filterConfigFilters , col)) {
+              defaultFilters[col] = filterConfigFilters[col];  
+            }
+          })
+  
+          // check date filter is applicable for filter
+          if (slice.form_data.date_filter) {
+            defaultFilters["__time_range"] = slice.form_data.time_range;
+          }
```
[the remark this answers is not in the document]
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/reducers/getInitialState.js]: Done

File: superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js (hunk 2 above)
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js]: done

File: superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js (hunk 5 above)
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js]: done

File: superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js (hunk 4 above)
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/util/charts/getFormDataWithExtraFilters.js]: done

File: superset/assets/src/dashboard/util/charts/getEffectiveExtraFilters.js [hunk 9]
```diff
@@ -16,13 +16,41 @@
...
+const getOperatorOfColumn = (linkedSlicesExistInFilters, filteringSliceId, col) => {
+  let op = 'in';
+  let subscribe_columns;
+
+  linkedSlicesExistInFilters.forEach( slice => { 
+    const sliceId = Object.keys(slice) && Object.keys(slice)[0];
+    if (sliceId == filteringSliceId) {
+      subscribe_columns = slice[sliceId]
+    }
+  })
```
[the remark this answers is not in the document]
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/util/charts/getEffectiveExtraFilters.js]: done

File: superset/assets/src/dashboard/reducers/getInitialState.js [hunk 10]
```diff
@@ -67,34 +67,52 @@ export default function(bootstrapData) {
...
-        }
+  const getSliceData = (sliceId, slices) => {
+    let slice_data;
+    slices.forEach (slice => {
+      if(slice.slice_id == sliceId){
+        slice_data = slice;
+      }
+    })
+    return slice_data;
+  }
```
[the remark this answers is not in the document]
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/reducers/getInitialState.js]: done

File: superset/assets/src/dashboard/components/Dashboard.jsx [hunk 11]
```diff
@@ -249,16 +251,15 @@ class Dashboard extends React.PureComponent {
...
+    if (publishers[filterKey]) {
+      const linked_slices = publishers[filterKey].subcribers;
+      const key = parseInt(sliceId);
       if (linked_slices instanceof Array) {
-        keyExists = linked_slices.indexOf(key) != -1
+        keyExists = linked_slices.indexOf(sliceId) != -1
       } else {
-        keyExists = linked_slices === key
+        keyExists = linked_slices === sliceId
       }
```
[the remark this answers is not in the document]
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/components/Dashboard.jsx]: done

**Pull request merged**